File: skills/test-graph/test_graph/sources/DeleteCluster.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//SOURCES ../sdk/java/src/main/java/com/hayden/testgraphsdk/sdk/*.java

import java.util.List;
import java.util.Map;

import com.hayden.testgraphsdk.sdk.ClusterLifecycle;
import com.hayden.testgraphsdk.sdk.ClusterLifecyclePlan;
import com.hayden.testgraphsdk.sdk.Node;
import com.hayden.testgraphsdk.sdk.NodeResult;
import com.hayden.testgraphsdk.sdk.NodeSpec;

public class DeleteCluster {
    private record Target(String target, String backend) {}

    private static final List<Target> TARGETS = List.of(
            new Target("local-preview", "local"),
            new Target("local-github-action", "github-action"),
            new Target("aws-preview", "aws")
    );

    private static final NodeSpec SPEC = NodeSpec.of("tg6.lifecycle.java.delete-cluster")
            .kind(NodeSpec.Kind.ASSERTION)
            .tags("tg6", "environment", "lifecycle-template", "jbang")
            .output("Runtime", "string")
            .output("DeleteCases", "string");

    public static void main(String[] args) {
        Node.run(args, SPEC, ctx -> {
            boolean skipKeepsActive = true;
            boolean destroyRuns = true;
            int cases = 0;
            for (Target target : TARGETS) {
                ClusterLifecyclePlan skip = ClusterLifecycle.deleteCluster(target.target(), target.backend(), false);
                ClusterLifecyclePlan destroy = ClusterLifecycle.deleteCluster(target.target(), target.backend(), true);
                cases += 2;
                skipKeepsActive &= !skip.shouldRun()
                        && "destroy-not-requested".equals(skip.skipReason())
                        && "kept-active".equals(skip.expectedState());
                destroyRuns &= destroy.shouldRun()
                        && "destroy".equals(destroy.environmentAction())
                        && destroy.tofuCommand().equals(List.of("tofu", "destroy", "-auto-approve"))
                        && "destroyed".equals(destroy.expectedState());
            }

            ClusterLifecyclePlan sample = ClusterLifecycle.deleteCluster("local-preview", "local", false);
            NodeResult result = NodeResult.pass(ctx.nodeId())
                    .assertion("all_targets_covered", cases == TARGETS.size() * 2)
                    .assertion("destroy_skips_without_intent", skipKeepsActive)
                    .assertion("destroy_runs_with_intent", destroyRuns)
                    .assertion("aws_destroy_requires_explicit_selection", ClusterLifecycle.deleteCluster("aws-preview", "aws", true).requiresExplicitSelection())
                    .publish("Runtime", "jbang")
                    .publish("DeleteCases", Integer.toString(cases));
            for (Map.Entry<String, String> entry : sample.published().entrySet()) {
                result.publish("Delete" + entry.getKey(), entry.getValue());
            }
            return result;
        });
    }
}

File: skills/test-graph/test_graph/sources/delete_cluster.py
# /// script
# requires-python = ">=3.10"
# dependencies = ["testgraphsdk"]
#
# [tool.uv.sources]
# testgraphsdk = { path = "../sdk/python", editable = true }
# ///
from __future__ import annotations

from testgraphsdk import NodeResult, NodeSpec, delete_cluster, node


TARGETS = (
    ("local-preview", "local"),
    ("local-github-action", "github-action"),
    ("aws-preview", "aws"),
)

SPEC = (
    NodeSpec("tg6.lifecycle.python.delete-cluster")
    .kind("assertion")
    .tags("tg6", "environment", "lifecycle-template", "uv")
    .output("Runtime")
    .output("DeleteCases")
)


@node(SPEC)
def main(ctx):
    skipped = [delete_cluster(target, backend, destroy_requested=False) for target, backend in TARGETS]
    destroy = [delete_cluster(target, backend, destroy_requested=True) for target, backend in TARGETS]

    result = (
        NodeResult.pass_(ctx.node_id)
        .assertion("all_targets_covered", len(skipped) == len(TARGETS) and len(destroy) == len(TARGETS))
        .assertion("destroy_skips_without_intent", all(not plan.should_run for plan in skipped))
        .assertion("skip_reason_records_missing_destroy_intent", all(plan.skip_reason == "destroy-not-requested" for plan in skipped))
        .assertion("skip_destroy_keeps_environment_active", all(plan.expected_state == "kept-active" for plan in skipped))
        .assertion("destroy_runs_with_intent", all(plan.should_run for plan in destroy))
        .assertion("destroy_uses_destroy_action", all(plan.environment_action == "destroy" for plan in destroy))
        .assertion("destroy_uses_destroy_command", all(plan.tofu_command == ("tofu", "destroy", "-auto-approve") for plan in destroy))
        .assertion("destroy_expected_state_recorded", all(plan.expected_state == "destroyed" for plan in destroy))
        .assertion("aws_destroy_requires_explicit_selection", delete_cluster("aws-preview", "aws", destroy_requested=True).requires_explicit_selection)
        .publish("Runtime", "uv")
        .publish("DeleteCases", str(len(skipped) + len(destroy)))
    )
    for key, value in delete_cluster("local-preview", "local", destroy_requested=False).published().items():
        result.publish(f"Delete{key}", value)
    return result


if __name__ == "__main__":
    main()
